tests/test_TFNetworkLayer.py

```diff
@@ -150,6 +150,141 @@ def test_concat_sources_missing_dim():
+    src_data = Data(name='t', shape=(None, 32, 128), dtype='float32', auto_create_placeholders=True)
+    window_size = 5
+    slice_size = window_size - 1
+
+    rnd = np.random.RandomState(42)
+    input_data = rnd.rand(10, 11, 32, 128)
+    seq_lens = np.array([11, 11, 11, 11, 11, 11, 11, 11, 11, 11])
+
+    with tf.variable_scope("src"):
+      src = InternalLayer(name="src", network=net, out_type={"dim": 128,
+                                                             "shape": (None, 32, 128),
+                                                             "batch_dim_axis": 0,
+                                                             "time_dim_axis": 1,
+                                                             "feature_dim_axis": 3,
+                                                             "sparse": False
+                                                             })
+      src.output = src_data
+      feed_dict = {src_data.placeholder: input_data,
+                   src_data.size_placeholder[0]: seq_lens}
+    with tf.variable_scope("win"):
+      win = WindowLayer(name="win", network=net, window_size=window_size, sources=[src], padding="valid",
+                        output=WindowLayer.get_out_data_from_opts(name="win", window_size=window_size,
+                                                                  padding="valid", sources=[src], network=net))
+    with tf.variable_scope("red"):
+      red = ReduceLayer(name="red", network=net, mode="AVG", axes="s:1", keep_dims=False, sources=[win],
+                        output=ReduceLayer.get_out_data_from_opts(name="red", network=net, mode="AVG",
+                                                                  axes="s:1", keep_dims=False, sources=[win]))
+    with tf.variable_scope("slice"):
+      sl = SliceLayer(name="slice", axis="t", slice_start=slice_size // 2, slice_end=-(slice_size // 2),
+                      sources=[src], network=net,
+                      output=SliceLayer.get_out_data_from_opts(name="slice", axis="t", slice_start=slice_size // 2,
+                                                               slice_end=-(slice_size // 2), sources=[src],
+                                                               network=net))
+    with tf.variable_scope("combine"):
+      combine = CombineLayer(name="combine", network=net, kind="add", sources=[sl, red],
+                             output=CombineLayer.get_out_data_from_opts(name="combine", network=net,
+                                                                        kind="add", sources=[sl, red]))
+    tf.global_variables_initializer().run()
+    in1, in2 = session.run([combine.sources[0].output.placeholder,
+                            combine.sources[1].output.placeholder],
+                           feed_dict=feed_dict)
+    assert in1.shape == (10, 7, 32, 128)
+    assert in2.shape == (10, 7, 32, 128)
+
+    res = session.run(combine.output.placeholder, feed_dict=feed_dict)
```

Comment:
There's an error here:
      message = <local> u'Incompatible shapes: [10,7,32,1,128] vs. [10,1,7,32,128]\n\t [[node combine/Add (defined at /u/bozheniuk/crnn-dev/TFNetworkLayer.py:5612) ]]\n\nErrors may have originated from an input operation.\nInput Source operations connected to node combine/Add:\n combine/red_output_add_spatial_dim (defin..., len = 331
     InvalidArgumentError: Incompatible shapes: [10,7,32,1,128] vs. [10,1,7,32,128]

Previous assert are passed.

Reply:
Can you try to reproduce this problem just by using `Data` (or other things from `TFUtil`), and create a test case in `test_TFUtil.py`?

Reply:
I committed a new test in the test_TFUtit.py

Reply:
Once again, this is a test of a simplified topology:
                        1) ->  WindowLayer ->  ReduceLayer ->
InternalLayer                                                                          CombineLayer
                        2)         -> SliceLayer ->

In the example above, InternalLayer outputs the tensor of size [10, 11, 32, 128]
The branch 1) takes the output [10, 11, 32, 128] and first applies a window of size 5(the result is [10, 7, 5, 32, 128]) then a reduce layer (the result is [10, 7, 32, 128]). In general case, if the input is [b, t, s:1, f] the output of 1) is [b, t-(win_size-1), s:1,  f]
The branch 2) just slices the input with a SliceLayer (the result is [10, 7, 32, 128]). In general case, if the input is [b, t, s:1, f] the output of 1) is [b, t-slice_size, s:1,  f] = [b, t-(win_size-1), s:1,  f]
The CombineLayer tries to add the outputs from 1) and from 2), which have the same size. But fails.

Reply:
Is this related to the error you write? You get still `Incompatible shapes` here?

Reply:
You cannot know in general that branch 1 and branch 2 (and whatever other branch of modifying the size / variable-dim shape) are the same. How would you? (This is not just a rhetorical question. I don't really understand how you get that idea.)

So, what you want is, that you can explicitly mark in some way that they are in fact the same (that is the function `DimensionTag.declare_same_as`). Or, you could use `ReinterpretDataLayer` with `size_base`, like this:
```
"red1": {"class": "reinterpret_data", "from": "red", "size_base": "sl"},
"combine": {"class": "combine", "kind": "add", "from": ["sl", "red1"]},
```
That will reset the `size_placeholder`, and thus explicitly defines it to have the same time-dim.

Reply:
(Btw, better write the whole test case using a dictionary to define the layers. That's simpler than constructing it manually like you do.)

Reply:
yes, there's still an error: 
https://github.com/rwth-i6/returnn/pull/214#issuecomment-527103109

Reply:
I thought this discussion here was about `Incompatible shapes`, but now you write about a different issue.
Anyway, my suggestion was now specifically about what I think you actually want to have Although you still did not specifically said what you really want. I assume now that you want that the time dimensions (and their tags) are the same. (Please write explicitly if that is what you want.)

Reply:
I have two outputs, from different sources and I want them to be added together. If their tags have to be equal, than yes, I want the tags to be equal as well.

Reply:
They don't have to be equal. If they are not equal, i.e. you have two different time dims, you will just end up with two different time dims (as you would expect).
But as I understand you, you say, you want that they are considered as the same time dim.

Reply:
I wrote you above one way how you can get that. Did you try?

Reply:
> They don't have to be equal. If they are not equal, i.e. you have two different time dims, you will just end up with two different time dims (as you would expect).
> But as I understand you, you say, you want that they are considered as the same time dim.

Yes, I assure and consider, that they are the same.  

> I wrote you above one way how you can get that. Did you try?

You mean `ReinterpretDataLayer`? No, I didn't try

Reply:
Did you try now? Is this resolved?